### PR TITLE
Allow TableStyle to handle row headers

### PR DESCRIPTION
**Pull request by Xitian9**
This addresses some of the points in https://github.com/muesli4/table-layout/issues/20#issuecomment-1060164646. In particular, row headers are now rendered in the table, along with special line styles available within the row header (as there are in the column headers).

I have chosen to display a row header only in the first line of a row group, thus breaking the symmetry between row groups and row headers, so that question has also been resolved.

Also simplifies the creation of new TableStyle from TableStyleSpec, allowing it to handle ASCII tables as well, and creating an even simpler constructor simpleTableStyleSpec.

### CODE REVIEW

**Comment by Xitian9** [comment]
Here is an overly fancy example from the test suite. It renders better in my terminal.

```
╭───╥───────────────────┬───────────────────┬────────────┰────────────┬────────────┬────────────┬────────────┰────────────┬────────────╮
│   ║     Some text     │   Some numbers    │     X      ┃     Z      │     W      │     A      │     B      ┃    Text    │     Y      │
╞═══╬═══════════════════╪═══════════════════╪════════════╋════════════╧════════════╪════════════╧════════════╋════════════╧════════════╡
│ 1 ║ This is long text ┆    4.20000000     ┆    foo     ┃    blah         bloo    ┆    blop         blog    ┃   Short         baz     │
│   ║       Short       ┆ 200300400500600.2 ┆    bar     ┃   yadda         yoda    ┆   yeeda         york    ┃   Short        wibble   │
├───╫┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┼┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┼┄┄┄┄┄┄┄┄┄┄┄┄╂┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┼┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄╂┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┤
│ 2 ║ This is long text ┆    4.20000000     ┆    foo     ┃  bibbidy      babbidy   ┆    boo          blue    ┃   Short        wobble   │
╰───╨───────────────────┴───────────────────┴────────────┸─────────────────────────┴─────────────────────────┸─────────────────────────╯
```

**Comment by muesli4** [comment]
The end-result looks great, well done.  I will have a closer look soon.  (Just did some of the simple ones first.)

File: src/Text/Layout/Table.hs [hunk 1]
```diff
@@ -245,12 +244,12 @@ checkeredCells f g = zipWith altLines $ cycle [[f, g], [g, f]]
...
 -- specified for each column.
-colsG :: Monoid a => [Position V] -> [Col a] -> RowGroup a
+colsG :: Cell a => [Position V] -> [Col a] -> RowGroup a
 colsG ps = rowsG . colsAsRows ps
 
 -- | Create a 'RowGroup' by aligning the columns vertically. Each column uses
 -- the same vertical positioning.
-colsAllG :: Monoid a => Position V -> [Col a] -> RowGroup a
+colsAllG :: Cell a => Position V -> [Col a] -> RowGroup a
 colsAllG p = rowsG . colsAsRowsAll p
```
Comment (muesli4) [inline src/Text/Layout/Table.hs:267]: Why is this necessary?

**Comment by Xitian9** [comment]
Anything I can do to help this along?

File: src/Text/Layout/Table/Cell.hs [hunk 2]
```diff
@@ -37,17 +37,22 @@ class Cell a where
     -- the predicate matches.
     measureAlignment :: (Char -> Bool) -> a -> AlignInfo
 
+    -- | Create an empty cell.
+    -- This should satisfy `buildCell emptyCell = mempty`.
+    emptyCell :: a
```
Comment (muesli4) [inline src/Text/Layout/Table/Cell.hs:42]: I'm not sure if I asked this before but what is the point of it?  If this is just there to be used once in an internal algorithm then I have to question its existence.

File: src/Text/Layout/Table.hs (hunk 1 above)
Comment (Xitian9) [reply src/Text/Layout/Table.hs:267]: Not all things you would want to put into a table are instances of `Monoid`, and this would preclude their use with this function. In particular, the proposed `ElidableList` in one of the draft PRs does not have a `Monoid` instance.

The only thing the monoid constraint is needed for is padding, for which `mempty` is used. Under the assumption that one would only use this to later be rendered, we would already have a `Cell` instance and so may as well use the new `emptyCell` element of the typeclass.

That said, the most general solution may be to take the padding as an argument: `colsGWith :: a -> [Position V] -> [Col a] -> RowGroup a`. We could leave the existing `colsG = colsGWith mempty`, but I'm not sure this would be worth the effort.

File: src/Text/Layout/Table/Cell.hs (hunk 2 above)
Comment (Xitian9) [reply src/Text/Layout/Table/Cell.hs:42]: It's needed for padding, the same as the Monoid constraint on `colsG` and friends. Whenever we have more than one `Row` in a `RowGroup`, we display the row header only for the first row, and need to pad out the remaining rows with something. That is the role that `emptyCell` fills.

Without something to put there we would not be able to create row headers, at least not without doing something else, like repeating the row headers for every row in the group, which is not desired.

Comment (Xitian9) [reply src/Text/Layout/Table/Cell.hs:42]: Another thing to note: `Cell` is pretty much defined as being a typeclass which maps to various `Monoid`, though `buildCell` (since `StringBuilder` is a subclass of `Monoid`). In this context, `emptyCell` has the fairly natural and lawful definition of being an element which maps to `mempty` under `buildCell`.

File: src/Text/Layout/Table.hs (hunk 1 above)
Comment (Xitian9) [reply src/Text/Layout/Table.hs:267]: Just a follow-up that this ‘most general solution’ is actually implemented already as `vPad` and `vPadAll`. These functions are just convenience calls to that.

Comment (muesli4) [reply src/Text/Layout/Table.hs:267]: I am happy with your initial explanation already.

File: src/Text/Layout/Table/Cell.hs (hunk 2 above)
Comment (muesli4) [reply src/Text/Layout/Table/Cell.hs:42]: Thanks!

File: src/Text/Layout/Table/Cell/WideString.hs [hunk 3]
```diff
@@ -31,7 +31,7 @@ instance Cell WideString where
 -- The provided `Bool` determines whether to continue dropping zero-width
 -- characters after the requested width has been dropped.
 dropWide :: Bool -> Int -> String -> String
-dropWide _ i [] = []
+dropWide _ _ [] = []
```
Comment (muesli4) [inline src/Text/Layout/Table/Cell/WideString.hs:34]: Unrelated change.

File: src/Text/Layout/Table/LineStyle.hs [hunk 4]
```diff
@@ -100,13 +102,67 @@ asciiVertical NoLine     = ""
...
+-- Right joins
+arJoins DoubleLine NoLine     DoubleLine _          = "::"
+arJoins DoubleLine NoLine     _          _          = ":"
+arJoins SingleLine NoLine     DoubleLine _          = "++"
+arJoins SingleLine NoLine     _          _          = "+"
+-- Interior joins
+arJoins DoubleLine _          DoubleLine _          = "::"
+arJoins DoubleLine _          _          _          = ":"
+arJoins _          _          DoubleLine _          = "++"
+arJoins _          _          _          _          = "+"
```
Comment (muesli4) [inline src/Text/Layout/Table/LineStyle.hs]: If this can be a separate commit functionally then please make it one.

**Review by muesli4 — COMMENTED** [review]
In general, it seems that there are some unrelated changes.  Please separate as far as possible (i.e. without breaking functionality).  Enormous commits also slow down review considerably.

If I am completely wrong with my judgment then please let me know.

File: src/Text/Layout/Table/LineStyle.hs (hunk 4 above)
Comment (Xitian9) [reply src/Text/Layout/Table/LineStyle.hs]: I've separated this code in particular into a separate commit. It's not used at that point, but it doesn't affect compilation or functionality.

File: src/Text/Layout/Table/LineStyle.hs [hunk 5]
```diff
@@ -100,13 +102,67 @@ asciiVertical NoLine     = ""
...
+-- | ASCII representations for joins using pluses.
 asciiJoinString :: LineStyle -> LineStyle -> String
-asciiJoinString DoubleLine DoubleLine = "++"
-asciiJoinString _          DoubleLine = "++"
-asciiJoinString DoubleLine _          = ":"
-asciiJoinString _          NoLine     = "-"
-asciiJoinString _          _          = "+"
+asciiJoinString h v = asciiJoinString4 h h v v
+
+-- | Ascii representations for joins using rounded joins.
```
Comment (muesli4) [inline src/Text/Layout/Table/LineStyle.hs]: Please use the term `ASCII`.

File: src/Text/Layout/Table/Style.hs [hunk 6]
```diff
@@ -3,69 +3,139 @@
...
 
 import Text.Layout.Table.LineStyle
 
 -- | Specifies the different letters to construct the non-content structure of a
 -- table.
+--
+-- This is quite low-level and difficult to construct by hand. If you want to
+-- construct your own, you may wish to use the higher-level interface provided
+-- by (in increasing order of detail): 'simpleTableStyleSpec', 'TableStyleSpec',
+-- 'unicodeTableStyleFromSpec', 'asciiTableStyleFromSpec', 'tableStyleFromSpec'.
```
Comment (muesli4) [inline src/Text/Layout/Table/Style.hs]: Do bullet points make sense?

**Review by muesli4 — COMMENTED** [review]
Awesome.  I like how your abstraction simplifies the style definitions.

File: src/Text/Layout/Table/Style.hs (hunk 6 above)
Comment (Xitian9) [reply src/Text/Layout/Table/Style.hs]: Yes they do.
